Allow request_exec with async I/O and more

pbrezina opened the pull request:
This PR add calls to ``libssh`` ``request_exec()`` and relevant functions in order
to allow non-blocking async I/O. I need this to replace ``parallel-ssh`` which
unfortunately became a dead project, incompatible with Python 3.12+.

Additionally, it adds call to ``libssh`` ``request_send_signal()`` which allows to
send signals to the remote process if SSH server supports it.

I also fixed few bugs that I stumbled upon.


##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
$sbj.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
N/A

pbrezina committed:
Add low level call to request_exec and send_eof

To allow asynchronous handling of the execution.

pbrezina committed:
Add direct call to request_send_signal

OpenSSH now supports receiving signals for a long time. This is
useful when you need to communicate with a process via signals.

pbrezina committed:
Handle EOF in recv_nonblocking

If recv_nonblocking is called and the ssh command ended with EOF
and empty output, the following error is produced because nbytes ==
SSS_EOF == -127:

```
File "src/pylibsshext/channel.pyx", line 112, in pylibsshext.channel.Channel.read_nonblocking
SystemError: Negative size passed to PyBytes_FromStringAndSize
```

It is better to handle it gracefully.

pbrezina committed:
Add direct call to is_eof

When doing asynchronous I/O it is usefull to know when EOF is hit
after polling for data.

pbrezina committed:
Keep reference of Session in Channel

It may happen that a garbage collector collects Session object befor
unclosed Channel. Then Channel's destructor tries to close the channel,
libssh session is accessed but it is already freed, thus causing
segfault.

Keeping reference to the Session object makes sure that Session
is collected after Channel.

pbrezina committed:
docs: fix exec_command example

pbrezina committed:
Add changelog fragments

webknjaz commented on docs/changelog-fragments/576.bugfix.rst:
```diff
@@ -0,0 +1,6 @@
+pylibssh no longer crashes when received EOF or when channel is not explicitly
+closed -- by :user:`pbrezina`
```
```suggestion
closed -- by :user:`pbrezina`.
```

webknjaz commented on docs/changelog-fragments/576.bugfix.rst:
```diff
@@ -0,0 +1,6 @@
+pylibssh no longer crashes when received EOF or when channel is not explicitly
+closed -- by :user:`pbrezina`
+
+Previously, pylibssh crashed if channel.recv was called and libssh returned
```
```suggestion
Previously, pylibssh crashed if ``channel.recv()`` was called and ``libssh`` returned
```

webknjaz commented on docs/changelog-fragments/576.bugfix.rst:
```diff
@@ -0,0 +1,6 @@
+closed -- by :user:`pbrezina`
+
+Previously, pylibssh crashed if channel.recv was called and libssh returned
+SSH_EOF error. It also crashed on some special occasions where channel was not
```
```suggestion
the ``SSH_EOF`` error. It also crashed on some special occasions where channel was not
```

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -2,9 +2,13 @@
 import time
 
 import pytest
+from _service_utils import ensure_ssh_session_connected  # noqa: WPS436
```
Those utils aren't meant to be imported directly. They only serve as helpers to the fixtures defined in `conftest.py`.
Use `ssh_client_session` and you'll get the same thing with less code.

pbrezina replied:
I can't, using `ssh_client_session` will not trigger the condition in Garbage Collector.

pbrezina replied:
I converted it to a fixture on its own.

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +72,82 @@ def test_read_bulk_response(ssh_client_session):
+
+    # Without fix, garbage collector first deletes session and we segfault
+    # in channel destructor when trying to access low-level C session object.
+    _do_not_crash()
+    gc.collect(0)
```
It should be more stable if you first disable GC, then collect everything, and only then run the test helper, finally performing the collection call that checks for crashes. You could also assert the number of objects, maybe. Before and after. Then, re-enable GC upon completion of the test.

You could integrate `pypytools` and make a helper like https://github.com/cherrypy/cheroot/blob/0fd16f0c3f111c3709a4d14e6400cabf2c094239/cheroot/test/test_server.py#L412-L416 but that's not mandatory. Manual invocations in just one place is fine.

pbrezina replied:
I can't assert number of objects since it can be easily changed. Even `gc.is_tracked` can not be asserted as for some reason, it returns False for ssh_channel, but True for ssh_session. But I added `gc.disable` to have a better control over it.

Anyway, the crash is always triggered by the test code. Feel free to suggest improvements if you see any.

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +72,82 @@ def test_read_bulk_response(ssh_client_session):
+def test_is_eof(ssh_channel):
+    """Test that EOF-state is correctly obtained with is_eof."""
+    ssh_channel.request_exec('exit 0')
+    ssh_channel.poll(timeout=5000)  # noqa: WPS432
```
Disabling this linting rule is not justified here. Assign a name to that arbitrary number of `5_000` that labels it with some context.

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -170,9 +180,23 @@ cdef class Channel:
     def get_channel_exit_status(self):
         return libssh.ssh_channel_get_exit_status(self._libssh_channel)
 
+    def is_eof(self):
```
This method doesn't depend on any arguments and its name doesn't suggest it causes a side effect (it doesn't contain a verb). So it's actually a property, semantically:
```suggestion
    @property
    def is_eof(self):
```

webknjaz replied:
Could you also add a docstring for this?

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -170,9 +180,23 @@ cdef class Channel:
+        if rc != libssh.SSH_OK:
+            raise LibsshChannelException("Failed to ssh_channel_send_eof: [%d]" % rc)
+
+    def send_signal(self, sig):
```
Could you make this accept constants from the standard library module called `signal`?
Also, add a docstring, please.

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -170,9 +180,23 @@ cdef class Channel:
 
         return result
 
+    def send_eof(self):
```
Add a docstring here, please.

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -103,6 +109,10 @@ cdef class Channel:
             # This is what Session._get_session_error_str() does, but we don't have the Python object
             error = libssh.ssh_get_error(<void*>self._libssh_session).decode()
             raise LibsshChannelReadFailure(error)
+
+        if nbytes == libssh.SSH_EOF:
```
Perhaps, it's an `elif`?
```suggestion
        elif nbytes == libssh.SSH_EOF:
```

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -70,6 +71,11 @@ cdef class Channel:
         if rc != libssh.SSH_OK:
             raise LibsshChannelException("Failed to request_shell: [%d]" % rc)
 
+    def request_exec(self, command):
```
Could you add a docstring here?

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -46,6 +46,7 @@ cdef int _process_outputs(libssh.ssh_session session,
 
 cdef class Channel:
     def __cinit__(self, session):
+        self._session = session
```
@pbrezina is this the place with the actual fix for the SEGFAULT?

webknjaz replied:
Is this going to fix https://github.com/ansible/pylibssh/issues/57? If yes, then plz symlink `57.bugfix.rst` to `576.bugfix.rst`.

pbrezina replied:
It's likely, yes. But I can't tell without a coredump.

This is a fix for segfault I seen, test for it is in the test_destructor. The problem is that you open channel in cinit, then close it in destructor. To close it, you need to have valid session object. But if you rely on the destructor to close the channel, then it may happen that Session is gc'd first, therefore libssh_session is freed before you try to close the channel.

This patch makes sure that we keep reference to the python session object (not only to libssh session object) and therefore Channel is always gc'd before Session.

webknjaz replied:
Alright, thank you for the explanation! I'm marking this thread as resolved.

webknjaz commented on docs/changelog-fragments/576.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Published call to request_exec and other low level API -- by :user:`pbrezina`
```
```suggestion
The ``request_exec()`` method was added to the ``Channel``
class. It exposes an interface for calling the respective
low-level C-API of the underlying ``libssh`` library
-- by :user:`pbrezina`.
```

webknjaz commented on docs/changelog-fragments/576.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Published call to request_exec and other low level API -- by :user:`pbrezina`
+
+The following calls to libssh are now available: request_exec, send_eof, is_eof,
+request_send_signal.
```
```suggestion
Additionally, the following calls to ``libssh`` are now
available in the same class: ``request_exec()``, ``send_eof()``, ``request_send_signal()`` and ``is_eof``
which is exposed as a :py:class:`property`.
```

webknjaz commented on docs/changelog-fragments/576.doc.rst:
```diff
@@ -0,0 +1,3 @@
+Fixed example for exec_command -- by :user:`pbrezina`
+
+The example now contains correct method code and is fully functional.
```
```suggestion
Fixed the example of invoking remote commands by using
``Channel.exec_command()`` in snippets -- by :user:`pbrezina`.

Its previously showcased version wasn't as robust.
```

pbrezina replied:
I wouldn't say robust, it simply was not functional.

webknjaz replied:
@pbrezina was it nonfunctional unconditionally, then?

pbrezina replied:
Yes.

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +72,82 @@ def test_read_bulk_response(ssh_client_session):
+    def _do_not_crash():  # noqa: WPS430
+        ssh_session = Session()
+        ensure_ssh_session_connected(ssh_session, sshd_addr, ssh_clientkey_path)
+        ssh_channel = ssh_session.new_channel()  # noqa: F841
```
Would it be easier to trigger if this is linked to an external object?
```suggestion
        _do_not_crash.ssh_channel = ssh_session.new_channel()  # noqa: F841  # setting a non-accessed var is needed for testing GC
```

pbrezina replied:
I don't understand what you mean, can you ellaborate? But it is very easy to trigger it. Really just open session and channel and then let the program exit.

I added your suggestion.

webknjaz replied:
@pbrezina but in tests you have a separate nested function in hopes the variables in its scope would be garbage-collected in a breaking matter, right?

pbrezina replied:
Right. It needs to be garbage collected, so it needs to go out of scope.

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +72,82 @@ def test_read_bulk_response(ssh_client_session):
+
+    Test that this event does not cause a segfault in channels destructor.
+    """
+    def _do_not_crash():  # noqa: WPS430
```
```suggestion
    def _do_not_crash():  # noqa: WPS430  # required to create a garbage-collection scope
```

pbrezina commented:
It is ready for next round of review. I addresses all comments, leaving few as "unresolved" in case you want to discuss it further.

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -15,14 +15,17 @@
 from libc.string cimport memset
 
 from pylibsshext.errors cimport LibsshChannelException
+
```
Did some linter want these formatting changes?

pbrezina replied:
It must have been isort.

pbrezina replied:
Removed.

webknjaz commented on docs/changelog-fragments/576.bugfix.rst:
```diff
@@ -0,0 +1,7 @@
+pylibssh no longer crashes when received EOF or when channel is not explicitly
+closed -- by :user:`pbrezina`.
+
+Previously, pylibssh crashed if ``channel.recv`` was called and ``libssh``
```
I just realized that this refers to `pylibssh` which is not an identifier used anywhere in this project. The PyPI project is `ansible-pylibssh` and the corresponding GitHub slug is `ansible/pylibssh`. The importable is `pylibsshext`. Let's try using the project name here. I hope the `|project|` RST substitution declared in `docs/conf.py` will just work:

```suggestion
|project| no longer crashes when received EOF or when channel is not explicitly
closed -- by :user:`pbrezina`.

Previously, |project| crashed if ``channel.recv`` was called and ``libssh``
```

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -170,9 +183,35 @@ cdef class Channel:
+        if not isinstance(sig, signal.Signals):
+            raise TypeError(f"Expecting signal.Signals not {type(sig)}")
+
+        sshsig = sig.name.replace("SIG", "")
```
```suggestion
        sshsig = sig.name.replace("SIG", "")  # FIXME: replace w/ `str.removeprefix()` once Python 3.8 support is dropped
```

webknjaz commented on src/pylibsshext/channel.pyx:
```diff
@@ -170,9 +183,35 @@ cdef class Channel:
+        """
+        Send signal to the remote process.
+
+        @sig is a signal constant from ``signal``, e.g. ``signal.SIGUSR1``.
```
Let's follow the native Sphinx param list style just like in other places:
```suggestion
        :param sig: a signal constant from ``signal``, e.g. ``signal.SIGUSR1``.
        :type sig: 
```

Alternatively, you could've used double backticks for `sig`. But the `@sig` syntax is not something that is typical for Python (I believe it comes from Java or JS?).

pbrezina replied:
doxygen, fixed

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +74,87 @@ def test_read_bulk_response(ssh_client_session):
+    """
+    ssh_channel.request_exec('exit 0')
+    ssh_channel.poll(timeout=POLL_TIMEOUT)
+    ssh_channel.recv()
```
Does this test rely on the session reaching EOF before invoking `recv()`? If yes, I suggest being explicit about that in the test:
```suggestion
    assert ssh_channel.is_eof
    ssh_channel.recv()
```

webknjaz commented on tests/conftest.py:
```diff
@@ -122,6 +123,23 @@ def ssh_client_session(sshd_addr, ssh_clientkey_path):
 
 
+@pytest.fixture
+def ssh_session_connect(sshd_addr, ssh_clientkey_path):
```
Could you inject this fixture into `ssh_client_session` so there's no duplication?

pbrezina commented:
Thank you. It is ready for next round.

webknjaz reviewed (approved):
Thank you for the well-explained patch!

webknjaz commented on tests/unit/channel_test.py:
```diff
@@ -68,3 +74,88 @@ def test_read_bulk_response(ssh_client_session):
+    assert rc == 0
+
+
+def test_send_signal(ssh_channel):
```
@pbrezina any idea why this test might be SEGFAULTing sometimes? See the CI log: https://github.com/ansible/pylibssh/actions/runs/7798061967/job/21271285995?pr=576#step:15:319.

pbrezina replied:
If I read it correctly this test timed out when connecting to ssh server:
```
 >       raise TimeoutError('Timed out waiting for a successful connection')
E       TimeoutError: Timed out waiting for a successful connection
```

I saw this from time to time when running tests locally (also with the tests not related to my changes).

Segfault comes from
```
XFAIL tests/unit/channel_test.py::test_exec_command - reason: This test causes SEGFAULT, flakily. Ref: https://github.com/ansible/pylibssh/issues/57; pytest-forked reason: :-1: running the test CRASHED with signal 11
```

pbrezina replied:
Which also means that #57 is not fixed by this.

pbrezina replied:
```
        libssh.ssh_channel_send_eof(channel)
        result.returncode = libssh.ssh_channel_get_exit_status(channel)
        if channel is not NULL:
            libssh.ssh_channel_close(channel)
            libssh.ssh_channel_free(channel)
```
I think you need to poll for exit status since it may still return -1 and you close the channel. Perhaps there is some race condition when the channel is being closed but callback is still called. We could be sure if we have coredump available but this is my wild guess.

webknjaz replied:
Would you be able to take a look? I kinda don't want to make the CI even more flaky with this patch..

pbrezina replied:
It's not getting more flaky since the problems are already there :-)

Anyway, I believe I have fixed the timeout issue. See the last commit.

For the segfault... is it possible to get a coredump and full backtrace out of the tests? I never hit it locally and I don't think we can fix it without it. We can try to implement my proposal, but then again, if it was true, it should be possible to write a solid reproducer and I was not able to find one.

webknjaz replied:
Well, I googled and couldn't understand how to retrieve that core. The faulthandler is enabled and prints out some trace through `catchsegv`. I wish there was a way to put it into a GHA artifact, though.

Thanks for the patch!

pbrezina committed:
tests: fix random timeout when connecting to ssh

Previous implementation of wait_for_svc_ready_state was not robust
enough as it could return even if further ssh connections kept failing.
